### ISSUE TITLE
ModelsBuilder LiveDll Mode, doesn't generate a dll in the absence of one.

### ISSUE THREAD

**Issue by mistyn8** [umbraco/Umbraco-CMS#7339]
ModelsBuilder LiveDll Mode, doesn't generate a dll in the absence of one.

## Umbraco version

I am seeing this issue on Umbraco version: 8.4.0 (running 3 tier on local + dev/live on umbraco cloud) 

Reproduction
------------

Using LiveDll for DEV/Live on umbracoCloud so that deployed UDA's don't require a manual model rebuild.
This is in conjunction with Dll mode on Local - so that not all touches of doctypes cause rebuild and we get IntelliSense and a dependency in the sister Code project.

So the bin/Umbraco.Web.Published.Models.dll isn't included in the repo, as it should be being auto-generated on dev/live.

However if on the dev/live env the dll doesn't exist then LiveDll Mode doesn't instigate a check and create. So we need to manually push an initial version of the dll.

Is this intended functionality?

**Comment by nul800sebastiaan** [comment]
I don't know if this is intended. Your workaround works though?

We don't recommend using LiveDll mode for anything but local dev as far as I can remember. 

I'll have to ask around, am not a MB expert at all, :)

**Comment by callumbwhyte** [comment]
If I've understood this correctly, you can't generate a DLL using Models Builder LiveDLL mode on your dev or live environment but does successfully generate the DLL on your local environment - only when the DLL already exists is it possible to regenerate the DLL?

Tested Umbraco 8.4 and LiveDLL mode successfully generates a DLL on every traditional environment I tested. IMHO this screams of an issue with permissions on the Umbraco Cloud side, which perhaps make sense given you don't want random code to just be able to dump any old DLL in the bin directly on live?

I question your need for using LiveDLL mode anywhere other than locally... What's the reason for requiring the DLL to be auto-generated on dev/live?

Why isn't/can't the `Umbraco.Web.Published.Models.dll` file be included in the repo and deployed?

I personally can't see a use case for changing doctypes / code directly on the live env and why this DLL would need to refresh, could you clarify?

My recommended Models Builder setup (and I believe Umbraco's too) is to use `AppData` mode and compile your models locally, commit the resulting DLL, and push to dev/live. This still gives you Intellisense locally. If you need the [models in a separate DLL](https://www.dot-see.com/en/blog/how-to-use-modelsbuilder-in-a-multi-project-umbraco-solution/) this can be configured too.

Happy to hear your thoughts and see if there's a better solution :-)

**Comment by mistyn8** [comment]
@callumbwhyte Thanks for the reply, suggestions and recommendations. :-)

So rightly or wrongly here's a few thoughts that have fed into this on our side.

Our content admins and UI designers are working on the dev env on cloud, not local instances and have access to manipulate docTypes.
With Umbraco Cloud and Deploy, UDA's should be the driving force behind deployment of doctype updates/amends, pureLive and liveDll mean that models builder keeps in sync without intervention.

So that's what leads to the MB.dll shouldn't be included in the repo.

Now for dev's we'd switched to Dll mode in the local instance as the way to gain intellisence, still maintaining the build models from the backend, rather than having to build the solution. (dll so that app not rebuilt on every save of a doctype developers wanted to be in control over app recycling ;-) )

So liveDll setting came on dev/live as we can't have Dll mode local and PureLive on cloud as ended up with duplicate class errors (app_data and dll versions exist)..
Maybe there is a way to have dll mode local and PureLive on cloud but we couldn't find a solution there.

On the actual bug flagged, I've just tried again deleting Umbraco.Web.Published.Models.dll.. locally or in kudu on cloud.. recycling the app in either doesn't produce the dll being regenerated :-( Which I think you say does work for you? SO not sure what the issue with my instances are there. ;(

And as a final thought with umb 8.5 removing Dll/LiveDll modes will have to look at another approach anyway!

**Comment by callumbwhyte** [comment]
@mistyn8 Great insight, thanks! I now totally understand your use case and your dilemma.

To clarify I can successfully generate a DLL using `LiveDLL` mode by using the Models Builder dashboard or when saving, it doesn't generate automatically on app-start. By deleting the DLL (and it not being tracked in Git) it's effectively the same as it never existing, but I think the file can still be replaced (e.g. regenerated).

What I'm unsure about still is why you can't check in the generated DLL... I assume if someone changes a doctype on the dev environment (and the updated DLL is generated) this will be automatically committed to the Umbraco Cloud repo? Then on your local environment you can fetch this generated DLL via `git pull`, or push a new version of that DLL. If you have a rule in your `gitignore` to ignore `Umbraco.Web.Published.Models.dll` then it won't be committed on change...

All that said, you are right that as of Umbraco 8.5 DLL mode won't be supported. AFAIK `PureLive` meets your needs on dev/live, but not locally as your devs need intelisense.

Perhaps a hybrid of approaches is needed here - `PureLive` on Cloud, and `AppData` mode locally. You could achieve this via an environment specific [config transform](https://our.umbraco.com/documentation/Umbraco-Cloud/set-up/Config-Transforms/), and generate your models into a separate DLL (as per my previous link) but never commit it. Seems a little convoluted though...

**Comment by mistyn8** [comment]
So it's that create a new cloud env... error as no dll for the models as we don't generate if missing :-) or on deploy dev to live.

Not sure but going from memory, uda changes on dev only resulted in uda's in the commit not the dll (when we had dll included in the repo)
Also just found productivity suffered, either pushing or pulling resulting in the dll not matching the uda's for the environment so needing the extra build models. If only an easy way to pull local and then run a build models action?
Dll mode on live meant too many it's broken oh forgot I had to rebuild models (hence liveDLL mode)
Also cherry picking commits into live as hotfixes the Mb.dll can be out of sync... where as relying on UDA's means always in sync with the environment we're on?
Did look in some part to config transforms, which would seem to be the approach that would work as you say... not sure why we ended up with collisions from app_data and dll classes... prob just trying to rush prior to xmas..

@callumbwhyte thanks again for your insight!
